README.md
# Excel-Module-Logger
## How to use
A module for Microsoft Excel that allows logging to a file.

All you have to do is download the Logger.bas and then in Excel under Visual Basic, Right Click on Modules and select "Import File" Select Logger.bas from where you downloaded it.

See Customization part of the readme on how to change how the log file is set up to be more of how you want it.

You can now call the sub in the following way.
```visual-basic
Call WriteToLog("Action")
```
Where Action is replaced by what you want written.


For example if in one of your macros you want to display that the user selected a sheet you would have the following in the sheet code.

```visual-basic
Private Sub Worksheet_Activate()
Call WriteToLog(Currentsheet.Name & " was selected")
End Sub
```

This will write something like "Sheet1 was selected" in the action part of the log.

For a better example of what you can do with this kind of code, in realses I've included an excel file with several examples.


## The Log File

Depending some of the constants you can change will effect the name of the file, however for now the set up of the file itself is all the same.

Below is an example

```
Date           Time            User       | Action
-------------------------------------------------------------------------------------------------
2015-07-20     09:33:58     GamerMan7799 | Workbook Saved
2015-07-20     09:34:01     GamerMan7799 | Report Menu opened
2015-07-20     09:34:57     GamerMan7799 | Workbook Saved
2015-07-20     09:36:11     GamerMan7799 | Workbook Saved
2015-07-20     09:36:13     GamerMan7799 | Workbook Closed
```



## Customization

There are currently 7 constant that you can change to cause the log to be saved differently, they are:

**gBlLogAction** - If logging is currently enabled or not, simple True or False. Default : True

**strBaseLogFileName** - The name of the log file, is string. Default : log.txt

**gBlAppendDateToLog** - If the date is added to the start of the log file, usuful for sorting on a computer. Default : True

**strLogFolderPath** - The path to the folder where the log files will be saved. If it is left blank it will be saved to the same folder as the workbook. Default : ""

**mbytDateFormat** - The Format that the date will take in the the file name (is that is on) and in the log. Has the following options:
* 0 = YYYY-MM-DD (Ex 2015-07-22) DEFAULT
* 1 = MM-DD-YYYY (Ex 07-22-2015)
* 2 = DD-MM-YYYY (Ex 22-07-2015)
* 3 = MM-DD (EX 07-22)
* 4 = YYYYMMDD (Ex 20150722)

**mblnUseMilitaryTime** - If military time should be used. If true time AM/PM will not be included. 3:00 PM would become 15:00 instead. Default : True

**mbytTimeFormat** - The format that the time will take in the log file. Has the following options:
* 0 = HH:MM:SS (EX 15:24:23 or 3:24:23 PM) DEFAULT
* 1 = HH:MM (EX 15:24 or 3:24 PM)
* 2 = HHMMSS (EX 152423 or 32423 PM)
* 3 = HHMM (EX 1524 or 324 PM)





ChangeLog.md
#Change Log
All notable changes to this project will be documented here.
This project adheres to [Semantic Versioning](http://semver.org/).

## [1.1.0] - 2015-07-22
### Added
* The option to change the format of the date, incldues:
  * YYYY-MM-DD  
  * MM-DD-YYYY  
  * DD-MM-YYYY
  * MM-DD
  * YYYYMMDD
* The option to use military time or not
* The option to change the format of the time, includes:
  * HH:MM:SS
  * HH:MM
  * HHMMSS
  * HHMM
* Version infomation in the header of the module

## [1.0.0] - 2015-07-20
### Added
* Initial


Logger.bas
Attribute VB_Name = "Logger"

'---------------------------------------------------------------------------------------
' Module    : Logger
' Author    : Patrick Rye
' Date      : 7/22/2015
' Version	: 1.1.0
' Purpose   : Holds a procedure which will write stuff to a log.
'---------------------------------------------------------------------------------------
Option Explicit
'   ----------------------------------------------------------------------------
Private Const gBlLogAction As Boolean = True 'Whether all actions should be logged or not
Private Const strBaseLogFileName As String = "log.txt" 'The base name of the file
Private Const gBlAppendDateToLog As Boolean = True 'Whether the date should be added to the front of the log file or not.
Private Const mbytDateFormat as Byte = 0 'What format the date is is
'0 = YYYY-MM-DD (default)
'1 = MM-DD-YYYY
'2 = DD-MM-YYYY
'3 = MM-DD
'4 = YYYYMMDD
Private Const mbytTimeFormat as Byte = 0 'The format of the time
'0 = HH:MM:SS (default)
'1 = HH:MM
'2 = HHMMSS
'3 = HHMM
Private const mblnUseMilitaryTime as Boolean = True 'If military time should be used (15:00 instead of 3:00 PM)
Private Const strLogFolderPath As String = "" 'The file path for the folder the log is placed into, leave blank to save to same folder as excel file.
'   ----------------------------------------------------------------------------
'Use the line to call this sub, remove ' before and replace *Log* with what you want logged
'Call WriteToLog(*Log*)
'   ----------------------------------------------------------------------------
Public Sub WriteToLog(strLine As String)
Dim strLogFilePath As String
Dim strUserName As String
Dim fso As Object
Set fso = CreateObject("Scripting.FileSystemObject")
Dim oFile As Object
If gBlLogAction <> True Then GoTo EndLogger 'If logging is not set to true then go to the end of the function
Dim strDate As String
Dim strTime As String
strDate = DateMaker() 'Make the Date
strTime = TimeMaker() 'Make the time
strUserName = (Environ$("Username")) 'Get the user name from the environmental variables
'Make the File Path of the log file.
If strLogFolderPath <> "" Then
    strLogFilePath = strLogFolderPath
Else
    strLogFilePath = ThisWorkbook.Path
End If
If Right(strLogFilePath, 1) <> "\" Then strLogFilePath = strLogFilePath & "\"
If gBlAppendDateToLog Then strLogFilePath = strLogFilePath & strDate & " "
strLogFilePath = strLogFilePath & strBaseLogFileName
'Checks if the log file already exits or not.
If Dir(strLogFilePath) <> "" Then
    'File Does exist, open it for appending
    Set oFile = fso.OpenTextFile(strLogFilePath, 8, True)
Else
    'File does not exist, create it with headers
    Set oFile = fso.CreateTextFile(strLogFilePath)
    oFile.WriteLine "Date           Time            User       | Action"
    oFile.WriteLine "-------------------------------------------------------------------------------------------------"
End If
oFile.WriteLine strDate & "     " & strTime & "     " & strUserName & "   | " & strLine
oFile.Close
EndLogger:
Set fso = Nothing
Set oFile = Nothing
End Sub
'   ----------------------------------------------------------------------------
Private Function DateMaker() As String
'Writes the date in the selected format
Dim strDate As String
Dim strMonth as String
Dim strDay as String
'Add a 0 to the front of the month and day if its less than 10.
'Makes it look nicer.
If Month(Now) < 10 Then
	strMonth = "0" & Month(Now)
Else
	strMonth = Month(Now)
End If
If Day(Now) < 10 Then
    strDay = "0" & Day(Now)
Else
    strDay = Day(Now)
End If
Select Case mbytDateFormat
	Case 0
		strDate = Year(Now) & "-" & strMonth & "-" & strDay
		'Ex 2015-07-22
	Case 1
		strDate = strMonth & "-" & strDay & "-" & Year(Now)
		'Ex 07-22-2015
	Case 2
		strDate = strDay & "-" & strMonth & "-" & Year(Now)
		'Ex 22-07-2015
	case 3
		strDate = strMonth & "-" & strDay
		'Ex 07-22
	case 4
		strDate = Year(Now) & strMonth & strDay
		'Ex 20150722
	case else
		strDate = Year(Now) & "-" & strMonth & "-" & strDay
end select
DateMaker = strDate
End Function
'   ----------------------------------------------------------------------------
Private Function TimeMaker() As String
'Writes the time in the selected format
Dim strTime As String
Dim strHour as String
Dim strMinute as String
Dim strSecond as String
Dim strAMPM as String

if mblnUseMilitaryTime then
	if Hour(Now) < 10 then
		strHour = "0" & Hour(Now)
	else
		strHour = Hour(now)
	end if
else
	if Hour(now) > 12 then
		strAMPM = "PM"
		strHour = (Hour(Now) - 12)
	elseIf Hour(Now) = 12 then
		strAMPM = "PM"
		strHour = "12"
	else
		strAMPM = "AM"
		strHour = Hour(now)
	end if
end if	

If Minute(Now) < 10 Then
    strMinute = "0" & Minute(Now)
Else
    strMinute = Minute(Now)
End If
If Second(Now) < 10 Then
    strSecond = "0" & Second(now)
Else
    strSecond = Second(now)
End If
Select case mbytTimeFormat
	Case 0
		strTime = strHour & ":" & strMinute & ":" & strSecond 
		'EX 15:24:23 or 3:24:23 PM
	case 1
		strTime = strHour & ":" & strMinute
		'EX 15:24 or 3:24 PM
	case 2
		strTime = strHour & strMinute & strSecond
		'EX 152423 or 32423 PM
	case 3
		strTime = strHour & strMinute
		'EX 1524 or 324 PM
	Case Else
		strTime = strHour & ":" & strMinute & ":" & strSecond
end Select
If Not(mblnUseMilitaryTime) then strTime = strTime & " " & strAMPM
TimeMaker = strTime
End Function

